### PR TITLE
Adding instructions on how to do monthly package updates for all parts of our architecture

### PR DESCRIPTION
**Pull request by maniarathi**
Requesting reviews from the following people for the following things:

@bkmartinjr -- Can you please check the instructions for cellxgene, both frontend and backend?
@Bento007 -- Can you verify the backend requirements files for Portal backend please and make sure I haven't missed anything?
@tihuan -- Can you please verify the instructions for the Portal frontend?

Everyone else is more FYI and to make sure the instructions makes sense.

### CODE REVIEW

File: how-we-work/how-to-upgrade-dependencies.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+The instructions to upgrade the Portal backend are pretty much identical to the [Explorer CZI-hosted backend instructions](#Upgrading-cellxgene-Explorer-czi-hosted-backend). The following is the list of files to check:
+
+- `corpora-data-portal/requirements.txt`
+- `corpora-data-portl/backend/chalice/api_server/requirements.txt`
+- `corpora-data-portal/backend/chalice/cloudfront_invalidator/requirements.txt`
+- `corpora-data-portal/backend/chalice/upload_failures/requirements.txt`
+- `corpora-data-portal/Dockerfile`
+- `corpora-data-portal/Dockerfile.processing_image`
+- `corpora-data-portal/Dockerfile.upload_failures`
```
Comment (MDunitz) [inline how-we-work/how-to-upgrade-dependencies.md:75]:
```suggestion
- `corpora-data-portal/.happy/requirements.txt`
- `corpora-data-portal/Dockerfile.upload_failures`
```

File: how-we-work/how-to-upgrade-dependencies.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+[Here](https://github.com/chanzuckerberg/cellxgene/pull/2167/files) is a sample PR that upgraded several frontend npm dependencies.
+
+### Upgrading cellxgene Explorer CZI-hosted backend
+
+The pip package dependency files for the CZI-hosted backend for cellxgene Explorer can be found at `cellxgene/backend/czi-hosted/requirements.txt` as well as `single-cell-infra/terraform/modules/hosted-cellxgene/common/customize/requirements.txt`. These two files should be in sync except that the file in `single-cell-infra` should pin all packages to a single version and the file in `cellxgene` should use the `>=` sign for the same version numbers. This is so that when we are testing locally, we can easily find out if a new release of a package is causing compatibility issues. We do want to pin versions in the `single-cell-infra` repo because these are the versions installed during deployment and we don't want an unexpected incompatibility due to a new package release to cause a production deployment to fail.
```
Comment (bkmartinjr) [inline how-we-work/how-to-upgrade-dependencies.md:25]: > We do want to pin versions

I would go so far as to say:  "We MUST pin versions ...".  Ie, it isn't a suggestion.

File: how-we-work/how-to-upgrade-dependencies.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+To perform the upgrades you can make use of `pip-upgrader` to figure out which packages have recent updates:
+
+1. Install [pip-upgrader](https://pypi.org/project/pip-upgrader/): `pip install pip-upgrader`.
+
+2. Upgrade the packages based on the report and verify that the application still works (please fully deploy to a `dev` or `rdev` environment to double check everything).
+
+You can also make use of the [Snyk reports](https://app.snyk.io/org/cellxgene) to figure out which specific packages have security vulnerabilities as well as the [`safety` package](#Upgrade-cellxgene-Explorer-desktop-backend).
```
Comment (bkmartinjr) [inline how-we-work/how-to-upgrade-dependencies.md:33]: Note that `safety` evaluates the *currently installed* packages.  To use it you have to create a clean virtual env, and install from the requirements file -- then run safety.

File: how-we-work/how-to-upgrade-dependencies.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+As a general rule please do not directly import the latest version of the downstream package that contains the vulnerability. For example, if we import package `A` and package `A` depends on package `B` which depends on package `C` which contains a vulnerability, please do not directly import the latest version of package `C` directly into our package. If a newer version of package `A` exists where they have addressed the vulnerability, directly upgrade `A`. Otherwise, wait until there is a new one available.
+
+For an overview of current dependency issues we have in cellxgene Explorer and cellxgene Portal, you can check the two corresponding projects in [Snyk](https://app.snyk.io/org/cellxgene). Note that you should take the findings with a grain of salt as some findings may not have remediation and therefore may continue to exist after everything else has been upgraded. It's a good place to double check your work though.
```
Comment (bkmartinjr) [inline how-we-work/how-to-upgrade-dependencies.md:7]: you should also look at Dependabot alerts in github.  We should not clear them unless:
* they are fully resolved
* they are not relevant

File: how-we-work/how-to-upgrade-dependencies.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+2. Upgrade the packages based on the report and verify that the application still works (please fully deploy to a `dev` or `rdev` environment to double check everything).
+
+You can also make use of the [Snyk reports](https://app.snyk.io/org/cellxgene) to figure out which specific packages have security vulnerabilities as well as the [`safety` package](#Upgrade-cellxgene-Explorer-desktop-backend).
+
+Finally, besides the `pip` installed packages, please also take a look at the Dockerfile (in the top level directory `cellxgene/Dockerfile`) to ensure that there are no security vulnerabilities there. Snyk is your friend to figure out if there are and what the remediation process is.
+
+### Upgrading cellxgene Explorer desktop backend
+
+The policy we have to upgrade the backend of cellxgene Explorer _desktop_ is different than any of the other policies. Because many users may have different versions of the dependent `pip` packages already installed on their laptop, and because many users do not use a virtual environment, we need to be _as permissive as possible_ when defining what the allowable versions of each package are.
```
Comment (bkmartinjr) [inline how-we-work/how-to-upgrade-dependencies.md:39]: This prose is somewhat unclear. Are we "permissive" as in allowing versions that have vulnerabilities?  Or as in allowing any version that is reasonably secure?

One possible policy:
* minimum version that is functionally compatible and includes no medium or high security risks
* it is acceptable to exclude specific versions (eg, `foo>=1.9.0,!=2.0.0`) if that is simpler

File: how-we-work/how-to-upgrade-dependencies.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+To update the `pip` packages for cellxgene Explorer desktop backend, we recommend doing the following steps:
```
Comment (bkmartinjr) [inline how-we-work/how-to-upgrade-dependencies.md:41]: as a side note: I have a suggestion from InfoSec team that may simplify this, but it requires some experimentation to know if it will work.  I have not had time to do the work.  LMK if you want me to pass it off to someone else.

**Review by bkmartinjr — CHANGES_REQUESTED** [review]
a few places (noted inline) could use some clarification...

File: how-we-work/how-to-upgrade-dependencies.md (hunk 4 above)
Comment (tihuan) [reply how-we-work/how-to-upgrade-dependencies.md:7]: E.g., https://github.com/chanzuckerberg/corpora-data-portal/security/dependabot

File: how-we-work/how-to-upgrade-dependencies.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+### Upgrading cellxgene Explorer frontend
+
+There are two commands to run to figure out which npm packages to upgrade:
+
+`npm audit` -- this command generates a report of known vulnerabilities in dependencies described in `cellxgene/client/package.json`.
+
+`npm outdated` -- this command generates a report of all dependencies in `cellxgene/client/package.json` and whether there are newer versions available.
+
+Run these two commands and upgrade packages to the latest available. Please also make any necessary changes in the client code to ensure that that the frontend still works as expected and is not using outdated references. Finally, run `npm i` to install the packages and update the corresponding `package-lock.json`.
```
Comment (tihuan) [inline how-we-work/how-to-upgrade-dependencies.md:19]: I'm thinking upgrading a library might need a little more guideline for different levels of upgrade: major, minor, patch?

E.g.,
```md
...

`npm outdated` -- this command generates a report of all dependencies in `cellxgene/client/package.json` and whether there are newer versions available.

Semantic version: Major.Minor.Patch (e.g., v1.2.3 means Major 1, Minor 2, Patch 3)

For major upgrade: Check the package's changelog, especially the breaking changes section first to see if such upgrade is straight forward to complete without significant effort. If so, upgrade away and make sure the affected functionalities still work. If significant effort is required, please create a ticket, document the required steps for upgrade, and bring it to the team's attention for pointing and prioritization

For minor and patch upgrade: In theory there should not be any breaking changes, so just upgrading the packages should not break anything. HOWEVER, not all packages follow semantic versioning, so please notice what packages have been upgraded in `package.json` and test their related functionality accordingly to ensure everything is fine
```

Also we don't need to manually update `package-lock.json` since it's machine generated file 😄 !

**Review by tihuan — COMMENTED** [review]
Love the writeup! Thanks so much for doing this, @maniarathi 🙏 Just added some small comments for the FE part!

**Review by Bento007 — APPROVED** [review]

File: how-we-work/how-to-upgrade-dependencies.md (hunk 7 above)
Comment (seve) [reply how-we-work/how-to-upgrade-dependencies.md:19]: Do we want to recommend re-building the package lock on semvar major upgrades?

Also, we might want to shout out helper tools like [npm-check-updates](https://github.com/raineorshine/npm-check-updates) and [npm-check](https://github.com/dylang/npm-check)

**Review by seve — COMMENTED** [review]
One comment regarding package lock and tooling

File: how-we-work/how-to-upgrade-dependencies.md (hunk 7 above)
Comment (tihuan) [reply how-we-work/how-to-upgrade-dependencies.md:19]: ooooh TIL! `npm-check` seems to provide more functionality, should we recommend that over `npm-check-updates`?

By re-building `package-lock.json`, do you mean deleting it and re-creating it?

Comment (seve) [reply how-we-work/how-to-upgrade-dependencies.md:19]: `npm-check-updates` has the [doctor](https://github.com/raineorshine/npm-check-updates#doctor-mode) functionality which is super powerful for safely updating every package to latest

Yep, that's what I meant!

File: how-we-work/how-to-upgrade-dependencies.md (hunk 1 above)
Comment (maniarathi) [reply how-we-work/how-to-upgrade-dependencies.md:75]: Do we not use the Dockerfile.upload_failures at all?

File: how-we-work/how-to-upgrade-dependencies.md (hunk 6 above)
Comment (maniarathi) [reply how-we-work/how-to-upgrade-dependencies.md:49]: Closing this comment since we chatted offline about this and have an open issue. Since we don't have it yet, I'm excluding prose about that option for now.

File: how-we-work/how-to-upgrade-dependencies.md (hunk 5 above)
Comment (maniarathi) [reply how-we-work/how-to-upgrade-dependencies.md]: Added more detailed prose. Your suggested policy is what I had in mind so pasted that in :)

File: how-we-work/how-to-upgrade-dependencies.md (hunk 3 above)
Comment (maniarathi) [reply how-we-work/how-to-upgrade-dependencies.md]: Added prose around this.

File: how-we-work/how-to-upgrade-dependencies.md (hunk 2 above)
Comment (maniarathi) [reply how-we-work/how-to-upgrade-dependencies.md]: Done

File: how-we-work/how-to-upgrade-dependencies.md (hunk 4 above)
Comment (maniarathi) [reply how-we-work/how-to-upgrade-dependencies.md:7]: Done

File: how-we-work/how-to-upgrade-dependencies.md (hunk 7 above)
Comment (tihuan) [reply how-we-work/how-to-upgrade-dependencies.md:21]: ooh I see! So we should definitely try `npm-check-updates`'s doctor command then! Thanks for the research 🚀 !

Comment (maniarathi) [reply how-we-work/how-to-upgrade-dependencies.md:21]: I'm going to make a decision here that we can revisit in the future: as part of monthly package updates, if there are breaking changes that occur due to major package updates, we should simply address them right away as part of housekeeping and ensuring that we don't continue to rack up debt. The failure I want to avoid is that we continuously deprioritize package updates because they are more work. Let's run into the case where the package update takes a ton of time away from a sprint, before installing the instructions to create a ticket and prioritize.

Comment (maniarathi) [reply how-we-work/how-to-upgrade-dependencies.md]: I updated the prose! Please re-review when you have a chance.

**Review by tihuan — APPROVED** [review]
NIIICE looks super complete to me! Thanks so much for adding this, @maniarathi  🎉 !!

File: how-we-work/how-to-upgrade-dependencies.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
+# How to Upgrade Dependencies
+
+Once a month at the beginning of the month, the oncall engineer should visit the current versions of package dependencies that we have for both the frontend (`npm`) and the backend (`pip`) for `corpora-data-portal` and `cellxgene`. We do this so that we keep up to date with any security vulnerabilities that may exist in older versions and so that we don't end up too out-of-date with improvements that occur in our dependencies causing forward-incompatibility. The sections below describe the policies and instructions for each subset of our product suite.
+
+As a general rule please do not directly import the latest version of the downstream package that contains the vulnerability. For example, if we import package `A` and package `A` depends on package `B` which depends on package `C` which contains a vulnerability, please do not directly import the latest version of package `C` directly into our package. If a newer version of package `A` exists where they have addressed the vulnerability, directly upgrade `A`. Otherwise, wait until there is a new one available.
```
Comment (bkmartinjr) [inline how-we-work/how-to-upgrade-dependencies.md:5]: This begs a policy question: if the immediately upstream dependency does not address the vuln in a timely manner, how should the vuln be resolved?  Example:  package `C` above is abandon-ware.

In some cases, the right decision will be to migrate to a different (actively supported) dependency.  In others, it will be to just wait.  And in others, we can safely ignore it. I think these likely require a case-by-case decision. 

I suggest that _when_ this occurs, the on-call engineer should file an issue, making note of the problem, and starting to track the likelihood of resolution.  At some point, if these issues are old enough, we should make that case-dependent decision.

File: how-we-work/how-to-upgrade-dependencies.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+Run the two above commands and upgrade packages to the latest available (that do not contain a security vulnerability). Please double check that the frontend still works as expected and is not using outdated references. Finally, before committing the updates, please rebuild `package-lock.json` by first deleteing `package-lock.json` and `node_modules` and then running `npm i` to generate the new `package-lock.json`.
+
+[Here](https://github.com/chanzuckerberg/cellxgene/pull/2167/files) is a sample PR that upgraded several frontend npm dependencies.
+
+#### Other useful tools
+
+Besides `npm audit` and `npm outdated`, you may take a look at [`npm-check`](https://www.npmjs.com/package/npm-check) and [`npm-check-updates`](https://www.npmjs.com/package/npm-check-updates) as tools to help upgrade packages to the latest version.
+
+### Upgrading cellxgene Explorer CZI-hosted backend
```
Comment (bkmartinjr) [inline how-we-work/how-to-upgrade-dependencies.md:35]: Missing from this section is a policy around the choice of version.  I think you want to always run the most recent *compatible* version of a package.   This is implicit; suggest making it explicit.

File: how-we-work/how-to-upgrade-dependencies.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+For an overview of current dependency issues we have in cellxgene Explorer and cellxgene Portal, you can check the two corresponding projects in [Snyk](https://app.snyk.io/org/cellxgene). Note that you should take the findings with a grain of salt as some findings may not have remediation and therefore may continue to exist after everything else has been upgraded. It's a good place to double check your work though.
+
+Finally, please check both repos' Dependabot Alerts to ensure that all alerts have been addressed. Please do not clear the alerts unless they are either addressed (via a fully merged PR) or they are not relevant. This means that if one of our dependency packages is unable to be updated because of no available upgrade, leave the alert open so we know to update the package later once an update is available. The Portal dependabot alerts can be found [here](https://github.com/chanzuckerberg/corpora-data-portal/security/dependabot) and the Explorer dependabot alerts can be found [here](https://github.com/chanzuckerberg/cellxgene/security/dependabot).
+
+## Upgrading cellxgene Explorer
+
+### Upgrading cellxgene Explorer frontend
```
Comment (bkmartinjr) [inline how-we-work/how-to-upgrade-dependencies.md:13]: Need to make the update policy explicit -- the front-end should  use the latest *compatible* version of any dependency.  The exceptions are were we have not yet upgraded between major releases (eg, D3), and there we should be on the latest minor version (and eventually burn down the debt of the major update).

File: how-we-work/how-to-upgrade-dependencies.md [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+1. Install [safety](https://pypi.org/project/safety/): `pip install safety`.
+
+1. Manually install the minimum required version of all packages in `cellxgene/backend/server/requirements.txt` and `cellxgene/backend/server/requirements-prepare.txt`.
+
+1. Run `safety`: `python -m safety check --full-report`
+
+1. Update requirements as indicated, goto step 2 again.  Repeat until clean.
+
+1. Check that cellxgene still works :)
```
Comment (bkmartinjr) [inline how-we-work/how-to-upgrade-dependencies.md:71]: as a future enhancement, you might want to add a git workflow that specifically tests that cellxgene passes unit & smoke tests when the minimum versions are installed.  This enhancement is unrelated to this document, but would catch incompatibilities sooner....

**Review by bkmartinjr — APPROVED** [review]
approving, with some errata noted inline.

File: how-we-work/how-to-upgrade-dependencies.md (hunk 1 above)
Comment (MDunitz) [reply how-we-work/how-to-upgrade-dependencies.md:96]: Sorry was trying to add the happy requirements. Didnt mean to remove upload failures

**Review by MDunitz — APPROVED** [review]
lgtm

**Review by colinmegill — APPROVED** [review]

**Review by seve — APPROVED** [review]
